### PR TITLE
Task: Move Snackbar provider to Design System

### LINKED ISSUE #6713: Task: Snackbar code quality improvements
## Task Description
Per discussion: https://github.com/google/web-stories-wp/pull/6696

This is a placeholder to follow up on some tech debt regarding the redesign: 
- Pascal identified an area where we could share more code in the design system between the editor and the dashboard. 
- Samuel identified an improvement we could make to the Snackbar props API, renaming the `handle` to `on`

### PR DESCRIPTION
**Pull request by embarks**
## Context
- The editor and the dashboard were using a different `SnackbarProvider`. I implemented the design system's `Snackbar` in the Editor app and @swissspidy saw an opportunity to share code between the Editor and the Dashboard. 

## Summary
- There were differences in data structures between the two providers to contain notifications. I ended up using an array over an object with an id because it will always be short and it simplifies some of the logic in the provider and simplifies the hook's api. 
- I wanted to keep the concerns of these changes separate from the original PR (the base branch for this PR) because of the increased complexity and much larger review. 
- Most of the code changes should be deletions and file name changes. I'll go over some of the other changes in the next section and as comments in the review.

## Relevant Technical Choices
- There were snackbar containers as well as providers in each app, now the provider also provides the display container (as it did in the editor originally). 
- One of the changes related to the above-- a new file called `useApiAlerts` was added because the `SnackbarView` (now deleted) had effects in it to check on API errors. 
- There was also an effect in there to clear the errors if the path changed for the Dashboard. There's no router in the editor so I added that logic to the dashboard `AppContent` in `dashboard/app/index`

## User-facing changes
- There are a couple slight animation transition changes on the snackbar from moving some of the editor code into the design system but otherwise this is just a code change.

### QA

- [x] This is a non-user-facing change and requires no QA

### UAT

- [x] UAT should use the same steps as above.

### Does this PR have a security-related impact?
no

### Does this PR change what data or activity we track or use?
no

### Does this PR have a legal-related impact?
no

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testiing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #6713

### CODE REVIEW

File: assets/src/dashboard/app/index.js [hunk 1]
```diff
@@ -42,10 +44,10 @@ import {
 } from '../constants';
 
 import { AppFrame, LeftRail, NavProvider, PageContent } from '../components';
+import usePrevious from '../../edit-story/utils/usePrevious';
```
Comment (swissspidy) [inline assets/src/dashboard/app/index.js]: Let's move this shared hook to `design-system/utils` too.

**Review by samwhale — COMMENTED** [review]
wowz
image

File: assets/src/design-system/components/snackbar/snackbarProvider.js [hunk 2]
```diff
@@ -23,9 +23,9 @@ import { v4 as uuidv4 } from 'uuid';
 /**
```
Comment (samwhale) [inline assets/src/design-system/components/snackbar/snackbarProvider.js]: 💭 I feel like the `SnackbarProvider` and the `useSnackbar` hook shouldn't live in the `components` folder. Maybe they should live in their own folder in the design system like `contexts` or something 🤔. Not a big deal, but what do you think? Am I just being silly?

**Review by samwhale — APPROVED** [review]
This is super dope.

File: assets/src/design-system/components/snackbar/snackbarProvider.js (hunk 2 above)
Comment (BrittanyIRL) [reply assets/src/design-system/components/snackbar/snackbarProvider.js]: i really like the idea of separating the `SnackbarProvider` and `useSnackbar` out of components. +1 for `/contexts`.

**Review by BrittanyIRL — APPROVED** [review]
yes! this is sooooo much cleaner, thank you! 

one thing - Sam had mentioned only wanting 1 snackbar to show at once? I personally like having them stack, i think it makes more sense to see errors/actions at the time they occur as you have it currently.

Agree with @samwhale that separating out context from components in the design system would be nice.

**Comment by embarks** [comment]
> yes! this is sooooo much cleaner, thank you!
> 
> one thing - Sam had mentioned only wanting 1 snackbar to show at once? I personally like having them stack, i think it makes more sense to see errors/actions at the time they occur as you have it currently.
> 
> Agree with @samwhale that separating out context from components in the design system would be nice.

There is a sort of weird UX with this original implementation where the first message will "block" other messages from being seen but the timeouts remain the same. I think I may keep them as "stacking" and beg for forgiveness and understanding. I think we could improve various things about the visuals there, as well.

**Review by maxyinger — APPROVED** [review]
very clean. nice

**Comment by embarks** [comment]
> yes! this is sooooo much cleaner, thank you!
> 
> one thing - Sam had mentioned only wanting 1 snackbar to show at once? I personally like having them stack, i think it makes more sense to see errors/actions at the time they occur as you have it currently.
> 
> Agree with @samwhale that separating out context from components in the design system would be nice.

@BrittanyIRL I moved the code into `./contexts` and also added a `max` prop to the `snackbarContainer`, and some logic to smoothly remove them. Here's what it looks like with a few different timings and `max` values. The first one is how the dashboard looks on this branch now, as with the editor.

https://user-images.githubusercontent.com/41136059/111396744-f803d700-867c-11eb-9326-064b63507240.mp4

https://user-images.githubusercontent.com/41136059/111396748-f9350400-867c-11eb-887d-f13fa107ff62.mp4


:

https://user-images.githubusercontent.com/41136059/111396742-f508e680-867c-11eb-89a1-cbfc6d4398af.mp4

File: assets/src/edit-story/components/library/test/mediaEditDialog.js [hunk 3]
```diff
@@ -27,7 +27,7 @@ import MediaEditDialog from '../panes/media/local/mediaEditDialog';
 import { renderWithTheme } from '../../../testUtils';
 import ApiContext from '../../../app/api/context';
 import MediaContext from '../../../app/media/context';
-import SnackbarContext from '../../../../design-system/components/snackbar/context';
+import SnackbarContext from '../../../../design-system/contexts/snackbar/context';
```
Comment (BrittanyIRL) [inline assets/src/edit-story/components/library/test/mediaEditDialog.js]: nit: could just import `import {SnackbarContext} from '.././../../design-system/contexts`?

Comment (maxyinger) [reply assets/src/edit-story/components/library/test/mediaEditDialog.js]: Do we want to try and export everything directly from the root for the ds?

**Comment by embarks** [comment]
Since the redesign did not make the bug bash, I am going to merge this into the PR for #6521

**Pull request merged**